File: C-fast-float-opt-4x-arm-vectorized/fft.c
#include <math.h>
#include "fft.h"
#include <arm_neon.h>

// Internal variables
static int phasevec_exist = 0;
static FLOAT_VFO_TYPE phasevec[32][2];

// Public function
struct _complexblock *fft(int log2point,const struct _complexblock xy_in) {
    int i;
    static struct _complexblock xy_out;
    if (!phasevec_exist) {
	for (i=0; i<32; i++) {
	    int point = 2<<i;
	    phasevec[i][0] = cos(-2*M_PI/point);
	    phasevec[i][1] = sin(-2*M_PI/point);
	}
	phasevec_exist = 1;
    }
    for (i=0; i < (1<<log2point); i+=2) {
	unsigned int brev = i;
	brev = ((brev & 0xaaaaaaaa) >> 1) | ((brev & 0x55555555) << 1);
	brev = ((brev & 0xcccccccc) >> 2) | ((brev & 0x33333333) << 2);
	brev = ((brev & 0xf0f0f0f0) >> 4) | ((brev & 0x0f0f0f0f) << 4);
	brev = ((brev & 0xff00ff00) >> 8) | ((brev & 0x00ff00ff) << 8);
	brev = (brev >> 16) | (brev << 16);

	brev >>= 32-log2point;
	xy_out.re[brev] = xy_in.re[i];
	xy_out.im[brev] = xy_in.im[i];

	unsigned int brev2 = brev | (1<<(log2point-1));
	xy_out.re[brev2] = xy_in.re[i+1];
	xy_out.im[brev2] = xy_in.im[i+1];
    }

    // here begins the Danielson-Lanczos section
    int n = 1<<log2point;
    int l2pt=0;
    int mmax=1;


    l2pt++;
    for (int i=0; i < n; i += 2) {
//	COMPLEX_TYPE tempXY = xy_out[i+mmax]; // w_XY = 1
//	xy_out[i+mmax]  = xy_out[i] - tempXY;
//	xy_out[i     ] += tempXY;
	FLOAT_TYPE tempX = xy_out.re[i+mmax];
	FLOAT_TYPE tempY = xy_out.im[i+mmax];
	xy_out.re[i+mmax]  = xy_out.re[i] - tempX;
	xy_out.im[i+mmax]  = xy_out.im[i] - tempY;
	xy_out.re[i     ] += tempX;
	xy_out.im[i     ] += tempY;
    }
    mmax<<=1;

    FLOAT_VFO_TYPE w_X2 = phasevec[l2pt][0];
    FLOAT_VFO_TYPE w_Y2 = phasevec[l2pt][1]; l2pt++;
    for (int i=0; i < n; i += 4) {
//	COMPLEX_TYPE tempXY = xy_out[i+mmax]; // w_XY = 1
//	xy_out[i+mmax]  = xy_out[i] - tempXY;
//	xy_out[i     ] += tempXY;

//	COMPLEX_TYPE tempXY2 = w_XY2 * xy_out[i+1+mmax];
//	xy_out[i+1+mmax]  = xy_out[i+1] - tempXY2;
//	xy_out[i+1     ] += tempXY2;
	FLOAT_TYPE tempX = xy_out.re[i+mmax];
	FLOAT_TYPE tempY = xy_out.im[i+mmax];
	xy_out.re[i+mmax]  = xy_out.re[i] - tempX;
	xy_out.im[i+mmax]  = xy_out.im[i] - tempY;
	xy_out.re[i     ] += tempX;
	xy_out.im[i     ] += tempY;

	FLOAT_TYPE tempX2 = (FLOAT_TYPE)w_X2 * xy_out.re[i+1+mmax] - (FLOAT_TYPE)w_Y2 * xy_out.im[i+1+mmax];
	FLOAT_TYPE tempY2 = (FLOAT_TYPE)w_X2 * xy_out.im[i+1+mmax] + (FLOAT_TYPE)w_Y2 * xy_out.re[i+1+mmax];
	xy_out.re[i+1+mmax]  = xy_out.re[i+1] - tempX2;
	xy_out.im[i+1+mmax]  = xy_out.im[i+1] - tempY2;
	xy_out.re[i+1     ] += tempX2;
	xy_out.im[i+1     ] += tempY2;
    }
    mmax<<=1;

#ifdef MOD_SPEED
    while (l2pt < log2point) {
	int istep = 2<<l2pt;
#else
    while (n>mmax) {
	int istep = mmax<<1;
#endif
//	double theta = -2*M_PI/istep;
//	COMPLEX_TYPE wphase_XY = cos(theta) + sin(theta)*I;
	FLOAT_VFO_TYPE wphase_X = phasevec[l2pt][0];
	FLOAT_VFO_TYPE wphase_Y = phasevec[l2pt][1]; l2pt++;

	float32x4_t wphase_Xvec, wphase_Yvec;
	wphase_Xvec[0] = wphase_Xvec[1] = wphase_Xvec[2] = wphase_Xvec[3]= phasevec[l2pt+1][0];
	wphase_Yvec[0] = wphase_Yvec[1] = wphase_Yvec[2] = wphase_Yvec[3]= phasevec[l2pt+1][1];

	float32x4_t w_Xvec, w_Yvec;
	w_Xvec[0] = 1.;
	w_Yvec[0] = 0.;

	w_Xvec[1] = w_Xvec[0] * wphase_X; // - w_Yvec[0] * wphase_Y;
	w_Yvec[1] = w_Xvec[0] * wphase_Y; // + w_Yvec[0] * wphase_X;

	w_Xvec[2] = w_Xvec[1] * wphase_X - w_Yvec[1] * wphase_Y;
	w_Yvec[2] = w_Xvec[1] * wphase_Y + w_Yvec[1] * wphase_X;

	w_Xvec[3] = w_Xvec[2] * wphase_X - w_Yvec[2] * wphase_Y;
	w_Yvec[3] = w_Xvec[2] * wphase_Y + w_Yvec[2] * wphase_X;

	for (int m=0; m < mmax; m+=4) { // optimization: tempXY and tempXY2
	    for (int i=m; i < n; i += istep) {
/*		FLOAT_TYPE tempX = (FLOAT_TYPE)w_X * xy_out.re[i+mmax] - (FLOAT_TYPE)w_Y * xy_out.im[i+mmax];
		FLOAT_TYPE tempY = (FLOAT_TYPE)w_X * xy_out.im[i+mmax] + (FLOAT_TYPE)w_Y * xy_out.re[i+mmax];
		xy_out.re[i+mmax]  = xy_out.re[i] - tempX;
		xy_out.im[i+mmax]  = xy_out.im[i] - tempY;
		xy_out.re[i     ] += tempX;
		xy_out.im[i     ] += tempY;

		FLOAT_TYPE tempX2 = (FLOAT_TYPE)w_X2 * xy_out.re[i+1+mmax] - (FLOAT_TYPE)w_Y2 * xy_out.im[i+1+mmax];
		FLOAT_TYPE tempY2 = (FLOAT_TYPE)w_X2 * xy_out.im[i+1+mmax] + (FLOAT_TYPE)w_Y2 * xy_out.re[i+1+mmax];
		xy_out.re[i+1+mmax]  = xy_out.re[i+1] - tempX2;
		xy_out.im[i+1+mmax]  = xy_out.im[i+1] - tempY2;
		xy_out.re[i+1     ] += tempX2;
		xy_out.im[i+1     ] += tempY2;

		FLOAT_TYPE tempX3 = (FLOAT_TYPE)w_X3 * xy_out.re[i+2+mmax] - (FLOAT_TYPE)w_Y3 * xy_out.im[i+2+mmax];
		FLOAT_TYPE tempY3 = (FLOAT_TYPE)w_X3 * xy_out.im[i+2+mmax] + (FLOAT_TYPE)w_Y3 * xy_out.re[i+2+mmax];
		xy_out.re[i+2+mmax]  = xy_out.re[i+2] - tempX3;
		xy_out.im[i+2+mmax]  = xy_out.im[i+2] - tempY3;
		xy_out.re[i+2     ] += tempX3;
		xy_out.im[i+2     ] += tempY3;

		FLOAT_TYPE tempX4 = (FLOAT_TYPE)w_X4 * xy_out.re[i+3+mmax] - (FLOAT_TYPE)w_Y4 * xy_out.im[i+3+mmax];
		FLOAT_TYPE tempY4 = (FLOAT_TYPE)w_X4 * xy_out.im[i+3+mmax] + (FLOAT_TYPE)w_Y4 * xy_out.re[i+3+mmax];
		xy_out.re[i+3+mmax]  = xy_out.re[i+3] - tempX4;
		xy_out.im[i+3+mmax]  = xy_out.im[i+3] - tempY4;
		xy_out.re[i+3     ] += tempX4;
		xy_out.im[i+3     ] += tempY4;
*/
		float32x4_t *reg1_re = (float32x4_t *)&xy_out.re[i+mmax]; // 4 lanes reg
		float32x4_t *reg1_im = (float32x4_t *)&xy_out.im[i+mmax]; // 4 lanes reg

		float32x4_t temp_re = w_Xvec * *reg1_re - w_Yvec * *reg1_im; // 4 lanes mul
		float32x4_t temp_im = w_Xvec * *reg1_im + w_Yvec * *reg1_re; // 4 lanes mul

		float32x4_t *reg2_re = (float32x4_t *)&xy_out.re[i]; // 4 lanes reg
		float32x4_t *reg2_im = (float32x4_t *)&xy_out.im[i]; // 4 lanes reg

		*reg1_re = *reg2_re - temp_re; // 4 lanes sub&store
		*reg1_im = *reg2_im - temp_im; // 4 lanes sub&store 
		*reg2_re += temp_re; // 4 lanes add&store
		*reg2_im += temp_im; // 4 lanes add&store
	    }
	    float32x4_t w_Xtmp;
	    w_Xtmp = w_Xvec * wphase_Xvec - w_Yvec * wphase_Yvec; // 4 lanes rotate
	    w_Yvec = w_Xvec * wphase_Yvec + w_Yvec * wphase_Xvec; // 4 lanes rotate
	    w_Xvec = w_Xtmp;
	}
	mmax=istep;
    }
    return &xy_out;
}

File: Rust-tests/Rust-fast-float-opt-4x-vectorized-obsoleted/src/fft.rs
#![feature(stdsimd)]
use std::simd::{f32x4};

type FloatVfoType = f32;
type FloatType = f32;
type VectorType = f32x4;
// Internal variables
pub struct Fft {
    phasevec: (Vec<f32>, Vec<f32>) // = [[0.0; 2]; 32];
}

// Public function
impl Fft {
    pub fn new() -> Fft {
	let mut f = Fft {phasevec: (Vec::new(), Vec::new())};
	for i in 0..32 {
	    let point: i32 = 2<<i;
	    f.phasevec.0.push( (-2.0*std::f64::consts::PI/(point as f64)).cos() as FloatVfoType);
	    f.phasevec.1.push( (-2.0*std::f64::consts::PI/(point as f64)).sin() as FloatVfoType);
	}
	return f;
    }

    pub fn fft(&self, log2point: u32, xy_in: &(Vec<FloatType>, Vec<FloatType>)) -> (Vec<FloatType>, Vec<FloatType>) {
	let mut i=0;
	let mut xy_out: (Vec<FloatType>, Vec<FloatType>) = (vec![0.; 1<<log2point], vec![0.; 1<<log2point]);
	while i < 1<<log2point {
	    let mut brev: usize = i;
	    brev = ((brev & 0xaaaaaaaa) >> 1) | ((brev & 0x55555555) << 1);
	    brev = ((brev & 0xcccccccc) >> 2) | ((brev & 0x33333333) << 2);
	    brev = ((brev & 0xf0f0f0f0) >> 4) | ((brev & 0x0f0f0f0f) << 4);
	    brev = ((brev & 0xff00ff00) >> 8) | ((brev & 0x00ff00ff) << 8);
	    brev = (brev >> 16) | (brev << 16);

	    brev >>= 32-log2point;
	    xy_out.0[brev] = xy_in.0[i];
	    xy_out.1[brev] = xy_in.1[i];

	    let brev2 = brev | (1<<(log2point-1));
	    xy_out.0[brev2] = xy_in.0[i+1];
	    xy_out.1[brev2] = xy_in.1[i+1];
	    i+=2;
	}

	// here begins the Danielson-Lanczos section;
	let n = 1<<log2point;
	let mut l2pt=0;
	let mut mmax=1;


	l2pt+=1;

	i=0;
	while i < 1<<log2point {
	    let tempx: FloatType = xy_out.0[i+mmax];
	    let tempy: FloatType = xy_out.1[i+mmax];
	    xy_out.0[i+mmax]  = xy_out.0[i] - tempx;
	    xy_out.1[i+mmax]  = xy_out.1[i] - tempy;
	    xy_out.0[i     ] += tempx;
	    xy_out.1[i     ] += tempy;
	    i+=2;
	}
	mmax<<=1;

	let w_x2: FloatVfoType = self.phasevec.0[l2pt];
	let w_y2: FloatVfoType = self.phasevec.1[l2pt]; l2pt+=1;

	i=0;
	while i < n {
	    let tempx: FloatType = xy_out.0[i+mmax];
	    let tempy: FloatType = xy_out.1[i+mmax];
	    xy_out.0[i+mmax]  = xy_out.0[i] - tempx;
	    xy_out.1[i+mmax]  = xy_out.1[i] - tempy;
	    xy_out.0[i     ] += tempx;
	    xy_out.1[i     ] += tempy;

	    let tempx2: FloatType = w_x2 as FloatType * xy_out.0[i+1+mmax] - w_y2 as FloatType * xy_out.1[i+1+mmax];
	    let tempy2: FloatType = w_x2 as FloatType * xy_out.1[i+1+mmax] + w_y2 as FloatType * xy_out.0[i+1+mmax];
	    xy_out.0[i+1+mmax]  = xy_out.0[i+1] - tempx2;
	    xy_out.1[i+1+mmax]  = xy_out.1[i+1] - tempy2;
	    xy_out.0[i+1     ] += tempx2;
	    xy_out.1[i+1     ] += tempy2;
	    i+=4;
	}
	mmax<<=1;

	while n>mmax {
	    let istep = mmax<<1;
	    let wphase_x: FloatVfoType = self.phasevec.0[l2pt];
	    let wphase_y: FloatVfoType = self.phasevec.1[l2pt];

	    let wphase_xvec: VectorType = VectorType::splat(self.phasevec.0[l2pt-2]);
	    let wphase_yvec: VectorType = VectorType::splat(self.phasevec.1[l2pt-2]);
	    l2pt+=1;


	    let w_x0 = 1.;
	    let w_y0 = 0.;

	    let w_x1 = w_x0 * wphase_x; // - w_Yvec[0] * wphase_Y;
	    let w_y1 = w_x0 * wphase_y; // + w_Yvec[0] * wphase_X;

	    let w_x2 = w_x1 * wphase_x - w_y1 * wphase_y;
	    let w_y2 = w_x1 * wphase_y + w_y1 * wphase_x;

	    let w_x3 = w_x2 * wphase_x - w_y2 * wphase_y;
	    let w_y3 = w_x2 * wphase_y + w_y2 * wphase_x;

	    let mut w_xvec = VectorType::new(w_x0, w_x1, w_x2, w_x3);
	    let mut w_yvec = VectorType::new(w_y0, w_y1, w_y2, w_y3);


	    let mut m=0;
	    while  m < mmax { // optimization: tempXY and tempXY2
		let mut i=m;
		while i < n {
		    let reg1_re = VectorType::load_unaligned(&xy_out.0[i+mmax..i+mmax+4]); // 4 lanes reg
		    let reg1_im = VectorType::load_unaligned(&xy_out.1[i+mmax..i+mmax+4]); // 4 lanes reg

		    let temp_re: VectorType = w_xvec * reg1_re - w_yvec * reg1_im; // 4 lanes mul
		    let temp_im: VectorType = w_xvec * reg1_im + w_yvec * reg1_re; // 4 lanes mul

		    let reg2_re = VectorType::load_unaligned(&xy_out.0[i..i+4]); // 4 lanes reg
		    let reg2_im = VectorType::load_unaligned(&xy_out.1[i..i+4]); // 4 lanes reg

		    (reg2_re - temp_re).store_unaligned(&mut xy_out.0[i+mmax..i+mmax+4]); // 4 lanes sub&store
		    (reg2_im - temp_im).store_unaligned(&mut xy_out.1[i+mmax..i+mmax+4]); // 4 lanes sub&store 
		    (reg2_re + temp_re).store_unaligned(&mut xy_out.0[i..i+4]); // 4 lanes add&store
		    (reg2_im + temp_im).store_unaligned(&mut xy_out.1[i..i+4]); // 4 lanes add&store

		    i+=istep;
		}
		let w_xtmp: VectorType = w_xvec * wphase_xvec - w_yvec * wphase_yvec; // 4 lanes rotate
		w_yvec = w_xvec * wphase_yvec + w_yvec * wphase_xvec; // 4 lanes rotate
		w_xvec = w_xtmp;
		m+=4;
	    }
	    mmax=istep;
	}
	return xy_out;
    }
}
